finality-aleph/src/justification.rs

```diff
@@ -1,70 +1,121 @@
-use crate::{AuthorityId, AuthorityKeystore, AuthoritySignature, JustificationNotification};
+use crate::{
+    network, party::SessionInstance, AuthorityKeystore, JustificationNotification, KeyBox,
+    Signature,
+};
+use aleph_bft::{MultiKeychain, NodeIndex, SignatureSet};
 use codec::{Decode, Encode};
 use futures::channel::mpsc;
 use sp_api::{BlockT, NumberFor};
-use sp_application_crypto::RuntimeAppPublic;
 use sp_blockchain::Error;
+use std::{collections::HashMap, sync::Arc};
 use tokio::stream::StreamExt;
 
-#[derive(Clone, Encode, Decode, PartialEq, Eq, Debug)]
+#[derive(Clone, Encode, Decode, Debug)]
 pub struct AlephJustification {
-    pub(crate) signature: AuthoritySignature,
+    pub(crate) signature: SignatureSet<Signature>,
 }
 
 impl AlephJustification {
-    pub fn new<Block: BlockT>(auth_crypto_store: &AuthorityKeystore, hash: Block::Hash) -> Self {
-        Self {
-            signature: auth_crypto_store.sign(&hash.encode()[..]),
-        }
+    pub(crate) fn new<Block: BlockT>(signature: SignatureSet<Signature>) -> Self {
+        Self { signature }
     }
 
-    pub(crate) fn _decode_and_verify<Block: BlockT>(
+    pub(crate) fn decode_and_verify<
+        Block: BlockT,
+        MK: MultiKeychain<Signature = Signature, PartialMultisignature = SignatureSet<Signature>>,
+    >(
         justification: &[u8],
         block_hash: Block::Hash,
-        authorities: &[AuthorityId],
-        number: NumberFor<Block>,
+        multi_keychain: &MK,
     ) -> Result<AlephJustification, Error> {
         let aleph_justification = AlephJustification::decode(&mut &*justification)
             .map_err(|_| Error::JustificationDecode)?;
 
-        let encoded_hash = &block_hash.encode()[..];
-        for x in authorities.iter() {
-            if x.verify(&encoded_hash, &aleph_justification.signature) {
-                return Ok(aleph_justification);
-            };
+        let valid =
+            multi_keychain.is_complete(&block_hash.encode()[..], &aleph_justification.signature);
+
+        if !valid {
+            log::debug!(target: "afa", "Bad justification decoded for block hash #{:?}", block_hash);
+            return Err(Error::BadJustification("Invalid justification".into()));
         }
 
-        log::debug!(target: "afg", "Bad justification decoded for block number #{:?}", number);
-        Err(Error::BadJustification(String::from(
-            "No known AuthorityId was used to sign justification",
-        )))
+        Ok(aleph_justification)
     }
 }
 
-// For now it is glorified proxy channel not doing anything useful
-pub struct JustificationHandler<Block: BlockT> {
+pub struct JustificationHandler<Block: BlockT, N: network::Network<Block> + 'static> {
     finalization_proposals_tx: mpsc::UnboundedSender<JustificationNotification<Block>>,
     justification_rx: mpsc::UnboundedReceiver<JustificationNotification<Block>>,
+    sessions: Arc<HashMap<u32, SessionInstance<Block>>>,
```

Comment:
I believe we should use `Arc<Mutex<_>>` for sessions here and everywhere else. Otherwise there is no way to modify the hashmap in `member.rs`.